File: whiteboard/app/src/main/java/com/ditto/whiteboard/transport/LiveFrameDispatcher.kt
package com.ditto.whiteboard.transport

import kotlin.coroutines.cancellation.CancellationException
import kotlinx.coroutines.channels.ReceiveChannel

/**
 * Drains lossy live frames without letting one closed recipient permanently kill the shared
 * dispatcher. Cancellation still propagates so transport shutdown remains prompt.
 */
internal suspend fun <Frame, Recipient> dispatchLiveFrames(
  frames: ReceiveChannel<Frame>,
  encode: (Frame) -> ByteArray,
  recipients: () -> Collection<Recipient>,
  maxSendSize: (Recipient) -> Long,
  send: (Recipient, ByteArray) -> Unit,
  onSent: (Recipient) -> Unit,
  onFailure: (Recipient?, Throwable) -> Unit,
  throttle: suspend () -> Unit,
) {
  for (frame in frames) {
    try {
      val bytes = encode(frame)
      recipients().forEach { recipient ->
        try {
          if (bytes.size.toLong() <= maxSendSize(recipient)) {
            send(recipient, bytes)
            onSent(recipient)
          }
        } catch (cancelled: CancellationException) {
          throw cancelled
        } catch (error: Exception) {
          onFailure(recipient, error)
        }
      }
    } catch (cancelled: CancellationException) {
      throw cancelled
    } catch (error: Exception) {
      onFailure(null, error)
    }
    throttle()
  }
}

File: whiteboard-apple/Packages/WhiteboardCore/Sources/WhiteboardCore/Transport/LiveFrameDispatcher.swift
import Foundation

/// Drains lossy live frames without letting one closed recipient permanently kill the shared
/// dispatcher. Cancellation still propagates so transport shutdown remains prompt.
public func dispatchLiveFrames<Frame, Recipient, Frames: AsyncSequence>(
  frames: Frames,
  encode: (Frame) throws -> Data,
  recipients: () -> some Collection<Recipient>,
  maxSendSize: (Recipient) throws -> Int64,
  send: (Recipient, Data) throws -> Void,
  onSent: (Recipient) -> Void,
  onFailure: (Recipient?, any Error) -> Void,
  throttle: () async -> Void
) async throws where Frames.Element == Frame {
  for try await frame in frames {
    do {
      let bytes = try encode(frame)
      for recipient in recipients() {
        do {
          let sizeLimit = try maxSendSize(recipient)
          if Int64(bytes.count) <= sizeLimit {
            try send(recipient, bytes)
            onSent(recipient)
          }
        } catch is CancellationError {
          throw CancellationError()
        } catch {
          onFailure(recipient, error)
        }
      }
    } catch is CancellationError {
      throw CancellationError()
    } catch {
      onFailure(nil, error)
    }
    await throttle()
  }
}
